### PR TITLE
Some optimizations for set functions

### PR DESCRIPTION
**Pull request by borkdude**

### CODE REVIEW

**Comment by borkdude** [comment]
@PEZ can you review these changes? I think they are safe optimizations.

File: src/squint/set.js [hunk 1]
```diff
@@ -122,7 +122,7 @@ export function rename_keys(map, kmap) {
   return ks.reduce((m, k) => {
     const newKey = core.get(kmap, k);
     if (core.contains_QMARK_(map, k)) {
-      return core.assoc(m, newKey, core.get(map, k));
+      return core.assoc_BANG_(m, newKey, core.get(map, k));
```
Comment (PEZ) [inline src/squint/set.js:134]: So, because we do `core.dissoc()` we are operating on a copy, and thus we can bang on it?

Comment (borkdude) [reply src/squint/set.js:134]: Indeed

File: src/squint/set.js [hunk 2]
```diff
@@ -148,20 +148,22 @@ export function join(xrel, yrel, kmap) {
     if (core.seq(xrel) && core.seq(yrel)) {
       const ks = intersection(core.set(core.keys(core.first(xrel))), core.set(core.keys(core.first(yrel))));
       const [r, s] = core.count(xrel) <= core.count(yrel) ? [xrel, yrel] : [yrel, xrel];
-      const idx = core.group_by(core.juxt(...ks), r);
+      const select = core.juxt(...ks);
+      const idx = core.group_by(select, r);
       return core.reduce((ret, x) => {
-        const found = core.get(idx, core.juxt(...ks)(x));
-        return found ? core.reduce((acc, y) => core.conj(acc, core.merge(y, x)), ret, found) : ret;
+        const found = core.get(idx, select(x));
```
Comment (PEZ) [inline src/squint/set.js:163]: Sweet.

**Comment by PEZ** [comment]
Looks safer than the previous code. 😄

File: src/squint/set.js (hunk 1 above)
Comment (borkdude) [reply src/squint/set.js:134]: I guess this is a good point, we could optimize `dissoc` to return the original if the amount of keys is empty.
And then we would be mutating the original. I'll make that a tad safer.

File: src/squint/set.js [hunk 3]
```diff
@@ -128,13 +128,17 @@ export function select(pred, xset) {
 
 export function rename_keys(map, kmap) {
   const ks = core.keys(kmap);
+  let without = core.dissoc(map, ...ks);
+  if (without === map) {
+    without = core.__copy(map);
+  }
```
Comment (PEZ) [inline src/squint/set.js:134]: I was wondering how you would make it safer. Neat!

**Pull request merged**